components/butterfly/src/rumor.rs

```diff
@@ -408,17 +477,60 @@ impl<T> RumorStore<T> where T: Rumor
     /// We don't care if this repeats - it just needs to be unique for any given two states, which
     /// it will be.
     fn increment_update_counter(&self) { self.update_counter.fetch_add(1, Ordering::Relaxed); }
+
+    pub fn keys(&self) -> Vec<RumorKey> {
+        self.read_entries()
+            .values()
+            .flat_map(HashMap::values)
+            .map(RumorKey::from)
+            .collect()
+    }
+}
+
+impl<T> RumorStore<T> where T: Transient
+{
+    /// Find rumors in our rumor store that have expired.
+    fn partitioned<B>(&self, expiration_date: DateTime<Utc>) -> (B, B)
+        where B: Default + Extend<T>
+    {
+        self.read_entries()
+            .values()
+            .flat_map(HashMap::values)
+            .cloned()
+            .partition(|rumor| rumor.expiration().expired(expiration_date))
+    }
+
+    pub fn expired(&self, expiration_date: DateTime<Utc>) -> Vec<T> {
+        self.partitioned(expiration_date).0
+    }
+
+    pub fn live(&self, expiration_date: DateTime<Utc>) -> Vec<T> {
+        self.partitioned(expiration_date).1
+    }
```

Comment:
It looks like `expired` and `live` can both be private without anything breaking.

Reply:
Nice, I'll update that.